### PR TITLE
cgo,native: fix optimal work length return

### LINKED ISSUE #225: native: Dgeqrf does not return optimum lwork in work[0]

### PR DESCRIPTION
**Pull request by kortschak**
@vladimir-ch Please take a look.

Fixes #225.

### CODE REVIEW

File: native/dgeqrf.go [hunk 1]
```diff
@@ -14,7 +14,7 @@ import (
 // parameters at entry and exit.
 //
 // work is temporary storage, and lwork specifies the usable memory length.
-// At minimum, lwork >= m and this function will panic otherwise.
+// At minimum, lwork >= max(1, m) and this function will panic otherwise.
 // Dgeqrf is a blocked QR factorization, but the block size is limited
```
Comment (vladimir-ch) [inline native/dgeqrf.go]: `max(1,n)`

File: cgo/lapack.go [hunk 2]
```diff
@@ -894,18 +894,18 @@ func (impl Implementation) Dgeqr2(m, n int, a []float64, lda int, tau, work []fl
 //
 // The C interface does not support providing temporary storage. To provide compatibility
```
Comment (vladimir-ch) [inline cgo/lapack.go]: We should remove these comments, they are not true anymore.

File: native/dgeqrf.go [hunk 3]
```diff
@@ -25,14 +25,14 @@ func (impl Implementation) Dgeqrf(m, n int, a []float64, lda int, tau, work []fl
 	nb := impl.Ilaenv(1, "DGEQRF", " ", m, n, -1, -1)
 	lworkopt := n * max(nb, 1)
 	lworkopt = max(n, lworkopt)
+	if len(work) < max(1, lwork) {
```
Comment (vladimir-ch) [inline native/dgeqrf.go]: You could do this check even before calling Ilaenv.

File: cgo/lapack.go (hunk 2 above)
Comment (kortschak) [inline cgo/lapack.go]: Please file an issue.

File: native/dgeqrf.go (hunk 3 above)
Comment (kortschak) [inline native/dgeqrf.go]: Done.

I'm going to file an issue to consider moving checkMatrix calls higher universally as well. The decision may be not to, but I think it's something that should be discussed: #228.

File: native/dgeqrf.go (hunk 1 above)
Comment (kortschak) [inline native/dgeqrf.go]: Done. This required a moderate rewording since that would have been incorrect as well.

**Comment by kortschak** [comment]
PTAL

**Comment by vladimir-ch** [comment]
LGTM

**Pull request merged**